My nft screen

DarkRavenn opened the pull request:
Эпик "Профиль". Часть 2. 

[decompositionProfileScreen.md](https://github.com/DarkRavenn/FakeNFT/blob/profile-module/docs/decompositionProfileScreen.md)

архитектура - MVP 
верстка - код
[доска](https://github.com/users/DarkRavenn/projects/3/views/1)  

---

### План: 
### 2. Экран "Мои NFT" ≈ 5h30m - 6h30m
**2.1** Экран "Мои NFT" на мок данных **≈ 3h - 4h**
  * Создать модель NFT с полями: images, name, author, price, rating **15m**
  * Создать массив мок данных (mockNFTs). **30m**
  * Реализовать UITableView, создать кастомную ячейку с изображением, названием, автором, рейтингом и ценой. **1-2h**
  * Добавить кнопку сортировки и список вариантов сортировки. **30m**
  * Реализовать сортировку массива mockNFTs. **30m**
  * Обработать пустое состояние (отображение заглушки). **15m**

**2.2** Экран "Мои NFT" на данных с сервера **≈ 2h30m**
  * Запрос и парсинг списка NFT через NetworkClient. **1h**
  * Обновить UI после получения данных. **1h**
  * Обновить сортировку для данных с сервера. **30m**

---

### Фактически: 
### 2. Экран "Мои NFT" 7h28m
**2.1** Экран "Мои NFT" на мок данных **4h 8m**
  * Создать модель NFT с полями: images, name, author, price, rating **3m**
  * Создать массив мок данных (mockNFTs). **30m**
  * Реализовать UITableView, создать кастомную ячейку с изображением, названием, автором, рейтингом и ценой. **3h**
  * Добавить кнопку сортировки и список вариантов сортировки. **15m**
  * Реализовать сортировку массива mockNFTs. **10m**
  * Обработать пустое состояние (отображение заглушки). **10m**

**2.2** Экран "Мои NFT" на данных с сервера **3h 20m**
  * Запрос и парсинг списка NFT через NetworkClient. **1h**
  * Обновить UI после получения данных. **2h**
  * Обновить сортировку для данных с сервера. **20m**

DarkRavenn committed:
Создана модель MyNFT и массив mockNFTs

DarkRavenn committed:
Реализована UITableView на моковых данных с кастомной ячейкой

DarkRavenn committed:
Добавлена сортировака

DarkRavenn committed:
Подготовка к использованию сети

DarkRavenn committed:
Экран Мои NFT переписан для работы с сетью

DarkRavenn committed:
Рефакторинг под архитектуру MVP

DarkRavenn committed:
Рефакторинг: перемещение файлов

ikibodo commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTController.swift:
```diff
@@ -0,0 +1,218 @@
+    private func showSortMenu() {
+        let alertController = UIAlertController(title: "Сортировка", message: nil, preferredStyle: .actionSheet)
+        
+        let sortByPrice = UIAlertAction(title: "По цене" + (currentSortType == .price ? " 🔹" : ""), style: .default) { _ in
+            self.presenter?.sortNFTs(by: .price)
+            self.currentSortType = .price
+        }
+        
+        let sortByRating = UIAlertAction(title: "По рейтингу" + (currentSortType == .rating ? " 🔹" : ""), style: .default) { _ in
+            self.presenter?.sortNFTs(by: .rating)
+            self.currentSortType = .rating
+        }
+        
+        let sortByName = UIAlertAction(title: "По названию" + (currentSortType == .name ? " 🔹" : ""), style: .default) { _ in
```
Элемент 🔹 для отметки выбора сортировки выглядит удобно, но в фигме или в описании его нигде не нашла, поэтому сомневаюсь считается ли это улучшение соответствием дизайну или нет.

DarkRavenn replied:
Индикацию выбора по хорошему всегда надо отображать, единственное надо было бы по хорошему у дизайнера(в пачке уточнить как лучше выделить)

ikibodo commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTController.swift:
```diff
@@ -0,0 +1,218 @@
+        tableView.reloadData()
+    }
+    
+    func showError(message: String) {
```
Показ алерта пользователю в случае ошибки выглядит хорошо.

ikibodo reviewed (commented):
Вроде всё ок. Оставила комментарий с сомнением насчет дополнительного элемента, улучшающего функционал, но отсутствующего в описании и фигме.

Helvetii951 commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTCell.swift:
```diff
@@ -0,0 +1,242 @@
+        return stackView
+    }()
+    
+    private lazy var authorDetailsStackView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        stackView.axis = .horizontal
+        stackView.alignment = .center
+        stackView.spacing = 4
+        return stackView
+    }()
+    
+    private let nftImageView: UIImageView = {
+        let imageView = UIImageView()
+        imageView.contentMode = .scaleAspectFit
+        imageView.clipsToBounds = true
+        imageView.layer.cornerRadius = 12
+        imageView.heightAnchor.constraint(equalToConstant: 108).isActive = true
+        imageView.widthAnchor.constraint(equalToConstant: 108).isActive = true
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
```
Используются и ленивые, и неленивые свойства UI компонентов. Рекомендую всегда делать их ленивыми. В некоторых случаях мы сможем сэкономить память. Да и такой подход предпочитают в большинстве компаний.

Helvetii951 commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTPresenter.swift:
```diff
@@ -0,0 +1,100 @@
+        view?.showLoading(true)
+        var failedNftIds: [String] = []
+        var loadedNfts: [Nft] = []
+        let dispatchGroup = DispatchGroup()
```
`DispatchGroup()` 👍

Helvetii951 commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTPresenter.swift:
```diff
@@ -0,0 +1,100 @@
+            myNFTs.sort { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
+        }
+
+        UserDefaults.standard.set(type.rawValue, forKey: sortTypeKey)
```
Лучше создать отдельный класс для хранения флагов и настроек. С ростом проекта их будет все больше и больше. Будет удобно найти их в одном месте.

danaskaya commented on FakeNFT/Resources/Assets.xcassets/like.imageset/Contents.json:
```diff
@@ -0,0 +1,23 @@
+    {
+      "filename" : "Favourites icons.png",
+      "idiom" : "universal",
+      "scale" : "1x"
```
Картинки 1x можно не добавлять. Они нужны были для iPhone без retina экранов, а последний такой - iPhone 5. Он не поддерживает минимальную версию приложения, поэтому можно удалить, чтобы не занимать лишнюю память

danaskaya commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTCell.swift:
```diff
@@ -0,0 +1,242 @@
+        contentView.addSubview(mainStackView)
+        
+        mainContentStackView.addArrangedSubview(nameAndRatingStackView)
+        mainContentStackView.addArrangedSubview(priceStackView)
```
Можно объединить:
```suggestion
        [nameAndRatingStackView, priceStackView].forEach {
        mainContentStackView.addArrangedSubview($0)
        }
```

danaskaya commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTCell.swift:
```diff
@@ -0,0 +1,242 @@
+    
+    // MARK: - Actions
+    @objc private func heartButtonTapped() {
+        if heartButton.tintColor == UIColor.white {
```
Можно записать одной строчкой: 
```suggestion
     heartButton.tintColor = (heartButton.tintColor == UIColor.white) ? UIColor.red : UIColor.white
```

danaskaya commented on FakeNFT/Scenes /Profile/MyNFT/MyNFTController.swift:
```diff
@@ -0,0 +1,218 @@
+        alertController.addAction(sortByPrice)
+        alertController.addAction(sortByRating)
+        alertController.addAction(sortByName)
+        alertController.addAction(cancelAction)
```
Можно сделать код более компактным и читаемым, используя метод forEach

danaskaya reviewed (commented):
👍🏻

DarkRavenn committed:
Правки по ревью:

- Используются только ленивые свойства UI компонентов
- удалены 1х изображения
- Добавлено использование тернарного оператора
- Добавлено использование forEach для улучшения читаемости
- Добавлен отдельный класс для хранения флагов и настроек пользователя
